### PR TITLE
8290249: Vectorize signum on AArch64

### PR DESCRIPTION
**Pull request by Bhavana-Kilambi**
This patch auto-vectorizes Math.signum intrinsic for float and  double
types on aarch64 (Neon and SVE). On SVE supporting machines, if the
MaxVectorSize <=16 the Neon code would be emitted and if the
MaxVectorSize > 16, the SVE code for the intrinsic would be emitted.

Following is the performance data for the micro test here -
test/micro/org/openjdk/bench/vm/compiler/VectorSignum.java

```
Benchmark	                Size    A	B       C
VectorSignum.doubleSignum	256	1.79	1.70	3.18
VectorSignum.doubleSignum	512	1.86	1.73	3.69
VectorSignum.doubleSignum	1024	1.89	1.74	2.98
VectorSignum.doubleSignum	2048	1.92	1.75	3.04
VectorSignum.floatSignum	256	3.34	3.06	3.92
VectorSignum.floatSignum	512	3.63	3.22	5.27
VectorSignum.floatSignum	1024	3.76	3.35	4.77
VectorSignum.floatSignum	2048	3.85	3.47	5.59

```
A, B , C machine descriptions given below -
A : 128-bit Neon machine
B : 256-bit SVE machine
C : 512-bit SVE machine

The numbers in the table are the gain ratios between the runtime (ns/op)
of the scalar, non-vectorized intrinsic code and the vectorized version
of the intrinsic (this patch).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290249](https://bugs.openjdk.org/browse/JDK-8290249): Vectorize signum on AArch64

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Nick Gasson](https://openjdk.org/census#ngasson) (@nick-arm - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9807/head:pull/9807` \
`$ git checkout pull/9807`

Update a local copy of the PR: \
`$ git checkout pull/9807` \
`$ git pull https://git.openjdk.org/jdk pull/9807/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9807`

View PR using the GUI difftool: \
`$ git pr show -t 9807`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9807.diff">https://git.openjdk.org/jdk/pull/9807.diff</a>

</details>

### CODE REVIEW

**Comment by theRealAph** [comment]
Please do not commit this until [](https://github.com/openjdk/jdk/pull/9346) is in.

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1679]:
```suggestion
      sve_orr(vtmp, T, jlong_cast(1.0)); // OR it with +1 to make the final result +1 or -1 depending
```

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3197]: This should be handled by the "SVE Integer/Floating-Point Compare - Vectors" code.

**Comment by Bhavana-Kilambi** [comment]
Hello, thank you for reviewing my patch. I have made the changes as suggested and waiting for the refactoring patch to be merged. I will then change my *ad files accordingly and put another patch for review in this PR.

**Comment by theRealAph** [comment]
> Hello, thank you for reviewing my patch. I have made the changes as suggested and waiting for the refactoring patch to be merged. I will then change my *ad files accordingly and put another patch for review in this PR.

The change to assembler.hpp is still not done.

**Comment by Bhavana-Kilambi** [comment]
> > Hello, thank you for reviewing my patch. I have made the changes as suggested and waiting for the refactoring patch to be merged. I will then change my *ad files accordingly and put another patch for review in this PR.
> 
> The change to assembler.hpp is still not done.
I mean I will put up the entire patch (with changes to the assember_aarch64.hpp, c2_MacroAssembler_aarch64.cpp as suggested by you and also the *ad files) once the refactoring patch is merged. I will need to make changes to the *ad files once the refactoring patch is merged, so I plan to put them all together in a single patch. Apologies if I wasn't clear.

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3516]: Please wrap all of this in `#ifdef ASSERT`

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3530]: I suggest something like

```    bool is_absolute = op2 == 0b11;```
....
```
      case GE: cond_op = (op2 << 2) | (is_absolute ? 0b01 : 0b00); break;            \
```

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp
Comment (Bhavana-Kilambi) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3516]: Thank you for reviewing. Could you please clarify by what exactly you mean by "Please wrap all of this in #ifdef ASSERT"?  Do you mean squashing the if conditions with the asserts? The assert macro calls are already inside a "#define".

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3516]: That whole block is nothing but assertions. It'd be easier to read (and a little quicker to compile) if the block were surrounded by #ifdef ASSERT.

Comment (Bhavana-Kilambi) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3516]: I am sorry it's still a bit unclear to me. So we can't add the `#ifdef ASSERT` within a `#define` block. Did you mean something like this -  
```
#ifdef ASSERT
  #define INSN(…)
  assert()
  assert()
  ...
else
  #define INSN(…)
  // no asserts here
  ...
  ...
#undef INSN
#endif
```

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp:3516]: Ah, sorry. I made a mistake.

**Comment by Bhavana-Kilambi** [comment]
The builds on aarch64 have failed as I missed adding parantheses in the assembler.hpp file. Will update with the new patch shortly.

**Review by theRealAph — APPROVED** [review]

**Pull request closed without merging**
